// @eddeee888/gcg-typescript-resolver-files/src/generateResolverFiles/handleGraphQLUninionType.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.handleGraphQLUninionType = void 0;
const utils_1 = require("../utils");
const handleGraphQLUninionType = ({ fieldFilePath, resolverName, normalizedResolverName, resolversTypeMeta, moduleName, }, { result, config: { emitLegacyCommonJSImports } }) => {
    const variableStatement = `export const ${resolverName}: ${resolversTypeMeta.typeString} = { /* Implement ${resolverName} union logic here */ };`;
    result.files[fieldFilePath] = {
        __filetype: 'generalResolver',
        content: `
    ${(0, utils_1.printImportLine)({
            isTypeImport: true,
            module: resolversTypeMeta.module,
            namedImports: [resolversTypeMeta.typeNamedImport],
            emitLegacyCommonJSImports,
        })}
    ${variableStatement}`,
        mainImportIdentifier: resolverName,
        meta: {
            moduleName,
            normalizedResolverName,
            variableStatement,
        },
    };
};
exports.handleGraphQLUninionType = handleGraphQLUninionType;
